File: services/anomaly-detector/idempotency.go
package main

import (
	"context"
	"errors"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// anomalyDedupScope namespaces this service's claims in the shared
// idempotency_keys table (see migrations/000008_idempotency_keys.up.sql —
// "Used by any consumer/service that must guarantee 'process this
// event_id/request exactly once'", a table that existed but had no actual
// caller until this fix).
const anomalyDedupScope = "anomaly_detection"

// claimTelemetryEventOnce atomically claims a source telemetry event's
// EventID for anomaly-detection processing, using the same
// INSERT...ON CONFLICT DO NOTHING RETURNING pattern already used by
// alert-service (see createIfDue in store.go) for race-safe, at-least-once
// -delivery-tolerant dedup. Kafka redelivering telemetry.processed after a
// crash between detection and offset commit would otherwise re-run
// detection and publish a second AnomalyDetected (with a new AnomalyID) for
// the same underlying reading — this claim makes that a no-op instead.
// Returns true only the first time a given eventID is claimed.
func claimTelemetryEventOnce(ctx context.Context, pool *pgxpool.Pool, eventID string) (bool, error) {
	var id string
	err := pool.QueryRow(ctx,
		`INSERT INTO idempotency_keys (key, scope) VALUES ($1, $2) ON CONFLICT (scope, key) DO NOTHING RETURNING id`,
		eventID, anomalyDedupScope,
	).Scan(&id)
	if errors.Is(err, pgx.ErrNoRows) {
		return false, nil // already claimed by an earlier delivery of this event
	}
	if err != nil {
		return false, fmt.Errorf("claim event %q for anomaly detection: %w", eventID, err)
	}
	return true, nil
}

File: services/anomaly-detector/idempotency.py
"""Exactly-once anomaly *detection*, the Python port of idempotency.go.

Atomically claims a source telemetry event's event_id before any
detection work runs — not just before the anomaly publish — using the
same INSERT...ON CONFLICT DO NOTHING RETURNING pattern already used by
alert-service for race-safe, at-least-once-delivery-tolerant dedup. Kafka
redelivering telemetry.processed after a crash between detection and
offset commit would otherwise both (a) publish a second AnomalyDetected
for the same reading, and (b) fold the same reading a second time into
the EWMA statistical baseline and the Isolation Forest's training buffer
even when (a) is correctly suppressed — claiming this early makes the
whole redelivery a no-op, not just the publish. See the README's
"Delivery semantics" section for why this distinction matters under this
system's at-least-once delivery model.
"""

from __future__ import annotations

from psycopg_pool import ConnectionPool

# Namespaces this service's claims in the shared idempotency_keys table
# (see migrations/000008_idempotency_keys.up.sql) — used by any
# consumer/service that must guarantee "process this event_id/request
# exactly once".
ANOMALY_DEDUP_SCOPE = "anomaly_detection"


def claim_telemetry_event_once(pool: ConnectionPool, event_id: str) -> bool:
    """Returns True only the first time a given event_id is claimed."""
    with pool.connection() as conn:
        cur = conn.execute(
            "INSERT INTO idempotency_keys (key, scope) VALUES (%s, %s) "
            "ON CONFLICT (scope, key) DO NOTHING RETURNING id",
            (event_id, ANOMALY_DEDUP_SCOPE),
        )
        row = cur.fetchone()
        conn.commit()
        return row is not None
